### ISSUE TITLE
Random discussions

### ISSUE THREAD

**Issue by cutoken** [keyvank/femtoGPT#2]
Hi,
Assert fails in tensor/mod.rs when number of heads = number of layers. Not sure if it's not a good number but in nanogpt, equal values are supported - see the cpu section (apples to oranges so please ignore if it's not a good comparison):

https://github.com/karpathy/nanoGPT

**Comment by keyvank** [comment]
@cutoken Have you tried removing the `train_data/` folder before starting the training process on a different architecture?

**Comment by cutoken** [comment]
Oops. That seems to be the reason. 
Btw, the default learning rate given for Adam seems to be really low. With around 0.001 it seems to be converging faster. That's the beauty of this project though: to play around and see how the training and quality is effected.

**Comment by keyvank** [comment]
@cutoken Be careful! With a higher learning-rate, you may get trapped into a sub-optimal solution :)

**Comment by cutoken** [comment]
Okay one more question:
Let's say I'm training it on one dataset.txt. After some time I stop the training and switch it to another dataset.txt. Is it possible to port only the weights from previous training and start training on new data set? Currently it asserts in the same place as above. I'm guessing it's because more things than weights are getting stored to disk.

**Comment by keyvank** [comment]
@cutoken You mean using 8-bit floating-point instead of f32?

Wow I love that "Quack, quack." LOL :laughing:

Yes you can, but it's a bit tricky. You have to make sure the vocab and vocab-sizes of previous dataset and new dataset are equal, otherwise, you have to somehow make sure the char->int mappings are the same in both datasets. (And also change the shape of `token_embedding` tensor, which is tricky)

**Comment by cutoken** [comment]
Aah okay. Can you explain a little on embedding.dat, optimizer.dat, pos_embedding.dat and tensor_num.dat files a little. Where do the weights for the layers get stored and is vocab same as character set or different. Sorry too many naive questions. If there is some paper or article that makes sense to get the basic understanding, you drop that instead :)

**Comment by keyvank** [comment]
Right now, the implementation is not generic, but you can just find/replace f32 with other types :slightly_smiling_face: 

Watch this: https://www.youtube.com/watch?v=kCc8FmEb1nY

`embedding.dat` contains the token-embeddings
`pos_embedding.dat` contrains the positional embeddings
`optimizer.dat` contains the state of the Adam optimizer
`tensor_{tensor_id}.dat` contains weights, biases, ... of different operations (E.g Linear, LayerNorm, ...) in the network

**Comment by cutoken** [comment]
Got it. If I could get the weights alone loadable so that we can switch the model around different data sets, I'll raise a PR.

**Comment by cutoken** [comment]
@keyvank 
Solved the lack of interchangeability of weights between datasets issue by the simpler solution of limiting the vocab to a range of ascii characters. Unfortunately as you can guess, it would not only impose restrictions on datasets having other kinds of characters but also impose a penalty on simpler data sets. However it would let same model be used across similar data sets. So no PR is coming but I can raise one with a optional command line switch like --vocab=ascii if needed.

**Comment by keyvank** [comment]
@cutoken I just added a ASCII tokenizer! :)

**Comment by cutoken** [comment]
Awesome @keyvank !

**Comment by cutoken** [comment]
I'm trying to use the ascii tokenizer but may be a bit of implementation is missing. Shouldn't the new Tokenizer trait be used instead of SimpleTokenizer everywhere ?

**Comment by keyvank** [comment]
@cutoken It's not used by default. In `src/main.rs` you should replace `let tokenizer = ...` to `let tokenizer = AsciiTokenizer {}`

Btw, have you been able to get good results with a larger model?

**Comment by pavan-a-4915** [comment]
Understood. I haven't tried with larger models yet. I'm actually playing with even smaller ones than the default settings. The previous output I provided was with 4 layer 4 head network. Today I'm trying with 2 layer 2 head and 1 layer 1 head networks. Thanks to the ascii tokenizing I'm able to switch datasets easily. However what I would love to see is being able to save and load the weights, biases and other network representations more granular so that it is possible to visualize the changing weights or even transfer weights more easily between networks.

Were you able to get your larger network going ? I'm interested in seeing whether the output has improved and what it's producing :)

**Comment by cutoken** [comment]
Understood. I haven't tried with larger models yet. I'm actually playing with even smaller ones than the default settings. The previous output I provided was with 4 layer 4 head network. Today I'm trying with 2 layer 2 head and 1 layer 1 head networks. Thanks to the ascii tokenizing I'm able to switch datasets easily. However what I would love to see is being able to save and load the weights, biases and other network representations more granular so that it is possible to visualize the changing weights or even transfer weights more easily between networks.Were you able to get your larger network going ? I'm interested in seeing whether the output has improved and what it's producing :)

**Comment by keyvank** [comment]
@cutoken Unfortunately I'm away from my powerful machine so couldn't experiment much. But yeah, probably we should try bigger models before getting anything meaningful.

P.s. I just pushed an optimization that makes everything around x3-4 times faster.

**Comment by cutoken** [comment]
That is great news. I'm going to give it a whirl!

**Comment by cutoken** [comment]
btw, I notice that even with rayon while all my hyper-threads are engaged they are not hitting 100% usage. I tried increasing the threads but it made it slower. Not sure if it's just an artifact of hyper threading/display or some bug in rayon.

**Comment by keyvank** [comment]
@cutoken 
Could have different reasons. Maybe some parts of the code are still single-threaded. Maybe there is more memory-operations compared to computations, leading to an idle cpu.
#4 could be an important reason why the cpu is idle.

**Comment by cutoken** [comment]
This is after 9 hours of training. I think the words are starting to make sense. Model is 6L/6H. Loss is continuing to reduce nicely. As I mentioned earlier, I have been so far using much higher learning rate (0.001).  When the loss stops reducing I';; go with higher model size

```

What like but wore pad wo me che nogns yous dares,
As supt it nind bupart 'the reed:
And hils not es
```

**Comment by keyvank** [comment]
@cutoken 
Im getting convinced this is working!! Is the loss already under 2.0?

**Comment by cutoken** [comment]
The loss was hovering between 1.9 and 2.1 for above result. I stopped it after a couple of hours to try the nanogpt for comparison. Here are the observations:
1) They use larger embedding size but less layers as default for cpu: 4L/4H/128 Embedding degree
2) They are doing some learning rate varying magic with adam: The loss drops like a rock with each iteration (200 iteration loss is same our 2k iteration loss)
3)  they kept dropout as 0.0
4) each step takes around 500ms vs our 2kms which all things considered is not bad at all for rust version (they peg the cpu at 100%)

Going forward I'll use above params for testing. We can compare it with nanogpt on quality at same number of iterations so that we know the implementation is correct. From my points above I think there is something different in the optimizer part which we can focus on as I think femto is converging slower than the nano one :)

**Comment by keyvank** [comment]
@cutoken Yeah something is definitely wrong with the optimizer😨 Could you plz send me your training data when you got a chance?

keyvankambakhsh@gmail.com

**Comment by cutoken** [comment]
Ran it for 2 days or so. Unfortunately the loss hasn't fallen below 1.7 (it's below 1.8 consistently). Sample output:

```
Giring city. Wherer I is thou losan, You come.

SERIANA:
We is the kis o parst ict as lipp I'd ou nord thought Vearnce
Wiveie ve are minder. Wis nushe shall wome
Frtue quien of hou alok:' avre we year stake,
we ususore sch have prinut to our fileng
```

**Comment by keyvank** [comment]
@cutoken Looks impressive. Have you tried the new optimizations? We are like x10 faster now.

**Comment by cutoken** [comment]
Not yet. That's one of the reason wanted to restart the training. Hope you had a chance to look into the optimizer issues. Actually I noticed that removing the weight decay reduced the loss jittering. May be there is some issue on that part :sweat_smile:

**Comment by keyvank** [comment]
@cutoken Yes the speedup was really unexpected for me too!

So yeah, now we have to figure out why our optimizer steps are not very efficient. Interesting that weight-decay breaks things... Dropout + Weight-decay are used to fight against over-fitting, so I'm not sure if it's a good idea to remove them :(

**Comment by cutoken** [comment]
You are correct. Anyways despite jitter, the loss does reduce properly. I'm using learning rate decay like this:

```

    let mut nlr
         = self.learning_rate - ((self.learning_rate - self.min_lr_rate)
            *(curr_step as f32/self.lr_decay_iterations as f32));
```

Above change seem to be helping with the training. Overall this current version of femtogpt not only is fast but also feels 'correct' for lack of a better word. May be you ended up fixing some underlying issues indirectly as part of performance fixes. It definitely converges faster but I don't know if its due to the lr decay change above or something else.

**Comment by cutoken** [comment]
Got it. I'll run this till the loss doesn't drop and will update on what it produces.

**Comment by cutoken** [comment]
@keyvank , there seems to be a bug in loading where when doing ctrl+c and restarting training the loss actually increases when compared with saved weights. however the first inference in main.rs is accurate and so is the step loss for first step. can you have a look. Till this is fixed the previously trained weights will always be overwritten and we will lose the training. Below is the output:

```
Vocab-size: 65 unique characters
Number of parameters: 816705
Generating text:
Why the are tenjond.

MENENIUS:
Well, I the long stear, not be male upon thee,
'That prut tering. Wi
Starting the training loop... (This make take hours to converge! be patient!)

Saving the model...
Step: 0 Loss: 1.580483 (Elapsed: 482ms)
Step: 1 Loss: 1.7405107 (Elapsed: 434ms)
Step: 2 Loss: 2.4709346 (Elapsed: 434ms)
Step: 3 Loss: 2.9418354 (Elapsed: 443ms)
Step: 4 Loss: 2.9977975 (Elapsed: 440ms)
Step: 5 Loss: 3.786684 (Elapsed: 437ms)
Step: 6 Loss: 3.9728856 (Elapsed: 436ms)
Step: 7 Loss: 3.4472175 (Elapsed: 438ms)
Step: 8 Loss: 3.1553347 (Elapsed: 440ms)
Step: 9 Loss: 3.1498506 (Elapsed: 442ms)
Step: 10 Loss: 2.8847642 (Elapsed: 441ms)
Step: 11 Loss: 2.8868973 (Elapsed: 437ms)
Step: 12 Loss: 2.8345814 (Elapsed: 439ms)
Step: 13 Loss: 2.8081226 (Elapsed: 441ms)

```

Let me know if you need weights to debug. I have taken backup.

**Comment by cutoken** [comment]
hi @keyvank , there is something still wrong with the optimizer and loss function. Basicallt for a given loss %, the other models perform better than ours. I tried with gpt example from tch-rs and for the same loss % their version was giving much better output. So my guess is that we are some how calculating wrong loss which might also be affecting the gradients.

**Comment by keyvank** [comment]
@cutoken I found an error in the gradients. Not sure why it's happening. working on it

**Comment by keyvank** [comment]
@cutoken Mask-layer gradient was totally wrong. Just fixed it in the last commit :facepalm:

**Comment by cutoken** [comment]
Will test with latest version. May be we should have a test suite that test some parts of the code and important calculations ?

**Comment by keyvank** [comment]
Yes, I have to add some tests

**Comment by cutoken** [comment]
The new version looks better. I would say almost equivalent to the tch-rs version in quality at 1.4 loss range.

```
Adistition gone; true; schistoes for mine souls!
Before your home, bariechts should be
Carlam on that's a worf quirer of him so.
What look'd lack away more
To him foot; one hour fortious of saves:
Son;
'Tis all Earl mmistling me.

HARSARTIO:
Why, idless of my mocks fan that were percious.
Having I will thou should and the muour ne'er shor
To purple, when duke thy are out him.
But will bid you doth remember nature.
Even OF hencomey, carniffeit, I joy
Warming; my deed, but it on of mortard,

```

**Comment by keyvank** [comment]
@cutoken How long did it take for you to reach 1.4?

**Comment by cutoken** [comment]
With my lr decay changes (start lr as 0.001 spread across 50k steps), it took around 4.5k steps to get it. Around an hour on my machine. Config of model is 4H/4L/128ED with hidden = 0.

**Comment by keyvank** [comment]
@cutoken I just ran your lr-decay code for the first time. I have been wasting my time... this is weirdly fast. even faster than pytorch :|

**Comment by cutoken** [comment]
Ya I have no clue how it works but it is indeed fast!

**Comment by keyvank** [comment]
@cutoken I added your work into the main branch. (It allows fancier tuning too)

**Comment by cutoken** [comment]
Perfect. Can't wait to try the warm up thingy! :D

**Comment by cutoken** [comment]
Nice! May be we could implement bpe/use the corresponding crate for encoding @keyvank . Will let us get even better results and faster training times.

**Comment by pcranaway** [comment]
@cutoken I've implemented `Tokenizer`s for Sentencepiece and Hugging Face's tokenizers in [my fork](https://github.com/pcranaway/femtoGPT/tree/experimental/sentencepiece)

It's experimental and @keyvank most likely will want to have an in-house implementation of these tokenizers without using external libraries, though.

I'll try training using SentencePiece + your optimizations and will report back.

**Comment by keyvank** [comment]
@pcranaway i won’t insist writing our own tokenizer though. We can perhaps write our own sentencepiece model “reader” (probably just a hashmap?) smth called MappingTokenizer for example, which will be filled with sentencepiece or similar libraries

**Comment by pcranaway** [comment]
> @pcranaway i won’t insist writing our own tokenizer though. We can perhaps write our own sentencepiece model “reader” (probably just a hashmap?) smth called MappingTokenizer for example, which will be filled with sentencepiece or similar libraries

that's basically what I meant! we can surely implement a SentencePiece reader ourselves

**Comment by keyvank** [comment]
@pcranaway let’s do this!

**Comment by cutoken** [comment]
How is the training going @pcranaway ? Hope loss has come down and you are able to see some good outputs!

**Comment by keyvank** [comment]
@cutoken Haven't you seen the updated README yet?

**Comment by cutoken** [comment]
omg! that's insane! that too in just 10 hours. What model config were you using ? How much loss was it able to drop down to ?

**Comment by pcranaway** [comment]
> How is the training going @pcranaway ? Hope loss has come down and you are able to see some good outputs!

I got disappointed because it was all random words and stopped (also I've been training for a few days and my electricity bill will probably make me homeless)

**Comment by keyvank** [comment]
@cutoken Check out the model along with train_data on `reddit` branch

**Comment by cutoken** [comment]
@pcranaway  now I'm worried about my elec bill :D 
With a good model config and sentencepiece I'm sure this can be done quite easily! 

@keyvank I'll check it out. Would be nice if we can dump whole training into a single shareable file for inference and also a dedicated executable/command line on existing one for easy inference. With how fast this is training won't be surprised if people start using it to get minor real world tasks done!

**Comment by keyvank** [comment]
@cutoken That's a really great idea! We can even have a pre-trained model on english dataset, and just let users fine-tune it with their own dataset

**Comment by cutoken** [comment]
Ya exactly. Remember that there are barely any from scratch 'usable' models on this front in the wild. everything is derived from a certain grass eating animal! :D

**Comment by cutoken** [comment]
hi @keyvank ,
I'm trying to join in on the sentencepiece bandwagon but unfortunately when trying to modify the main branch with reddit branch changes, I get below exception. The reddit branch works fine though. This is with the shakespeare dataset (which works great in reddit branch):

```

Vocab-size: 128 unique characters
Number of parameters: 832896
Generating text:
thread 'main' panicked at 'assertion failed: ind < self.shape[0]', /home/gocnet/nn/femtoGPT/src/tensor/mod.rs:38:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:117:5
   3: femto_gpt::tensor::TensorOps::get
   4: femto_gpt::gpt::GPT<O,R>::infer
   5: femto_gpt::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

**Comment by keyvank** [comment]
@cutoken Can you run reddit branch properly? without integrating anything in main branch?

**Comment by cutoken** [comment]
Already done running reddit branch  (with shakespeare as I want everything to be comparable) :)  

This sentencepiece thingy is quite good. Loss takes a while to stabilize but the quality even at higher loss is way better (as expected - since char gpt have to get a lot more correct to look correct). After the above infer code modification, main branch also works fine.  I'm using 1L/1H layers to quickly compare things (lower models don't move past a certain loss value obviously)

I gave "He who" as input and look what I got:

```

He whose my such To more to Rome
He whose point, good a glass,
He whose have what I can country,
He whom, That we should in the grou
He whom inter me gentleman: And
He whose honour! Shall us upon th
He whose is a prove my grown,
He whom I have purse: Sir, I w
He whose not with not, And thou d
He wholess of the fair. DUKE VIN
He whose have dold, good inst
He whole god secress is my mothe
He whom the rest at make a place.
He wholess against be me most g
He whose shall see, give thee, si
He whose me stir. ANTIGONUS:
He whose your such beg, And been
He whose that you, Believe, that'

```

**Comment by keyvank** [comment]
@cutoken Did you use a new sentencepiece vocab ran on Shakespeare dataset? Or did you use the model for Reddit dataset? This is really good I like it:)

**Comment by cutoken** [comment]
Generated a new sentencepiece vocab on the shakespeare with vocab size as 128. We will have to think how to package sentencepiece with femto. While it is good, I hope it won't end up bringing python eco system horrors to femto which is simple to run so far with no dependencies. Probably shipping the whole lib would make sense.

**Comment by pcranaway** [comment]
@cutoken as you said that error was because some words which occurred in the dataset were not in the vocab

I'm working on a in-house sentencepiece implementation, maybe we can make it generate a vocab and train a sentencepiece model inside femto if it's fast enough

**Comment by cutoken** [comment]
hi @keyvank ,
    let's say I want to add a new decoder layer (the one that gets constructed as part of 0..num_layers loop) at run time after the gpt::new() call, how do I go about it ? As I understand you are just pushing the computations one by one incremented by tensorid so adding a layer at a later point of time will need incrementing the ids for the next layers as well (for example adding one more decoder layer along with all the sub layers like attention etc means incrementing the vocab out and other variables outside the for loop ?)

**Comment by keyvank** [comment]
Hey @cutoken, let's continue these kind of questions inside separate github issues, this will help us organize things. Could you make a copy?

**Comment by keyvank** [comment]
@cutoken 
Hey! We are kind of letting the model find the best word to vec encoding for the words. Words that have similar meaning will be closer in this vector space.

Learn more about this by researching on: “Embedding layers”

**Comment by keyvank** [comment]
@cutoken Each CPU thread is processing 1 instance, that's why I passed 1 here

**Comment by keyvank** [comment]
@cutoken 
We don’t do that here. We let the model discover a code for each position itself (Dynamically). While this is simpler, giving it a pattern like sine/cosine probably makes it faster, and reduces num-params.

**Comment by keyvank** [comment]
Actually let’s implement this!

**Comment by cutoken** [comment]
Once you do let me know. Always ready to test the new things! :D

**Comment by ULis3h** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
JKorey, @RaG-cell, @0xMario27, @Slavaa00, @nitinbhatia-dev, @ConnorBlockchain, @tranbamanh229289, @beaubutton, @pcaversaccio, @amiral-jion, @seanuefa, @ztaylor, @fifths, @aflesher, @ianhitlgang, @lwj010, @madsleejensen, @miguel-msa, @john-tam, @dmytrochumakov
